### PR TITLE
Fix duplication of global annotation strings in reverse translation

### PR DESCRIPTION
**Pull request by KornevNikita**
This is a patch to avoid duplication of global annotation strings in reverse
translation caused by the fact that translator translates each
UserSemantic decoration, even if a similar decoration with the same string
has also been translated.

### CODE REVIEW

**Comment by KornevNikita** [comment]
@MrSidims @AlexeySachkov @Fznamznon @mlychkov please take a look

File: lib/SPIRV/SPIRVReader.cpp [hunk 1]
```diff
@@ -3400,7 +3400,15 @@ void SPIRVToLLVM::transIntelFPGADecorations(SPIRVValue *BV, Value *V) {
     SmallString<256> AnnotStr;
     generateIntelFPGAAnnotation(BV, AnnotStr);
     if (!AnnotStr.empty()) {
-      auto *GS = Builder.CreateGlobalStringPtr(AnnotStr);
+      Constant *GS = nullptr;
+      std::string StringAnnotStr = AnnotStr.c_str();
+      static unordered_map<std::string, Constant *> Annotations;
```
Comment (Fznamznon) [inline lib/SPIRV/SPIRVReader.cpp:3405]: In general `static` variables are not appreciated (for example in LLVM coding guidelines https://llvm.org/docs/CodingStandards.html#do-not-use-static-constructors ), because order of their initialization is not defined by C++ standard. Can we for example make this map a field of `SPIRVReader`?

File: lib/SPIRV/SPIRVReader.cpp [hunk 2]
```diff
@@ -3400,7 +3400,15 @@ void SPIRVToLLVM::transIntelFPGADecorations(SPIRVValue *BV, Value *V) {
...
+      if (Annotations.find(StringAnnotStr) != Annotations.end())
+        GS = Annotations[StringAnnotStr];
+      else {
```
Comment (Fznamznon) [inline lib/SPIRV/SPIRVReader.cpp:3408]: Just to be consistent with the `else` block.
```suggestion
      if (Annotations.find(StringAnnotStr) != Annotations.end()) {
        GS = Annotations[StringAnnotStr];
      } else {
```

File: lib/SPIRV/SPIRVReader.cpp (hunk 1 above)
Comment (KornevNikita) [reply lib/SPIRV/SPIRVReader.cpp]: Check new commit please, This is what you're talking about?

File: lib/SPIRV/SPIRVReader.cpp [hunk 3]
```diff
@@ -3400,7 +3400,14 @@ void SPIRVToLLVM::transIntelFPGADecorations(SPIRVValue *BV, Value *V) {
...
     if (!AnnotStr.empty()) {
-      auto *GS = Builder.CreateGlobalStringPtr(AnnotStr);
+      Constant *GS = nullptr;
+      std::string StringAnnotStr = AnnotStr.c_str();
+      if (AnnotationsMap.find(StringAnnotStr) != AnnotationsMap.end()) {
+        GS = AnnotationsMap[StringAnnotStr];
+      } else {
+        GS = Builder.CreateGlobalStringPtr(AnnotStr);
+        AnnotationsMap.insert(std::make_pair(StringAnnotStr, GS));
+      }
```
Comment (mlychkov) [inline lib/SPIRV/SPIRVReader.cpp:3410]:
```suggestion
      std::string StringAnnotStr = AnnotStr.c_str();
      auto AnnotItr = AnnotationsMap.find(StringAnnotStr);
      if (AnnotItr != AnnotationsMap.end()) {
        GS = AnnotItr->second();
      } else {
        GS = Builder.CreateGlobalStringPtr(AnnotStr);
        AnnotationsMap.emplace(std::move(StringAnnotStr), GS);
      }
```

File: lib/SPIRV/SPIRVReader.cpp (hunk 1 above)
Comment (Fznamznon) [reply lib/SPIRV/SPIRVReader.cpp]: Yup

**Review by Fznamznon — APPROVED** [review]

File: lib/SPIRV/SPIRVReader.cpp [hunk 4]
```diff
@@ -3400,7 +3400,15 @@ void SPIRVToLLVM::transIntelFPGADecorations(SPIRVValue *BV, Value *V) {
     SmallString<256> AnnotStr;
     generateIntelFPGAAnnotation(BV, AnnotStr);
     if (!AnnotStr.empty()) {
-      auto *GS = Builder.CreateGlobalStringPtr(AnnotStr);
+      Constant *GS = nullptr;
+      std::string StringAnnotStr = AnnotStr.c_str();
+      auto AnnotItr = AnnotationsMap.find(StringAnnotStr);
+      if (AnnotItr != AnnotationsMap.end()) {
+        GS = AnnotItr->second();
```
Comment (MrSidims) [inline lib/SPIRV/SPIRVReader.cpp:3407]:
```suggestion
        GS = AnnotItr->second;
```

**Review by MrSidims — APPROVED** [review]
Please fix the build error

File: lib/SPIRV/SPIRVReader.h [hunk 5]
```diff
@@ -159,6 +159,7 @@ class SPIRVToLLVM {
   SPIRVToLLVMPlaceholderMap PlaceholderMap;
   std::unique_ptr<SPIRVToLLVMDbgTran> DbgTran;
   std::vector<Constant *> GlobalAnnotations;
```
Comment (mlychkov) [inline lib/SPIRV/SPIRVReader.h:161]: It would be nice to have a comment about GlobalAnnotations and AnnotationsMap fields to understand what is the difference between them.

Comment (KornevNikita) [reply lib/SPIRV/SPIRVReader.h:161]: Done. Looks better now?

Comment (mlychkov) [reply lib/SPIRV/SPIRVReader.h:161]: Ok, let me rephrase this: May we join GlobalAnnotations and AnnotationsMap? And why if we cann't?

Comment (KornevNikita) [reply lib/SPIRV/SPIRVReader.h:161]: Do you mean to use only one container instead of these two?

Comment (mlychkov) [reply lib/SPIRV/SPIRVReader.h:161]: Yes.

File: lib/SPIRV/SPIRVReader.h [hunk 6]
```diff
@@ -159,6 +159,15 @@ class SPIRVToLLVM {
...
```
Comment (mlychkov) [inline lib/SPIRV/SPIRVReader.h:161]:
```suggestion
  // GlobalAnnotations vector collects structures of
  // annotations arguments for global variables and functions.
  // They are used in translation of llvm.global.annotations instruction.
  std::vector<Constant *> GlobalAnnotations;
```

File: lib/SPIRV/SPIRVReader.h [hunk 7]
```diff
@@ -159,6 +159,15 @@ class SPIRVToLLVM {
...
+  // There is also GlobalAnnotations vector which is used to store arrays of
+  // arguments for llvm.global.annotations instructions. But SPIR-V may also
+  // contain annotations for llvm.ptr.annotation.
+  // During reverse translation we need to check if a similar annotation with
+  // the same string has already been translated to avoid duplication of these
+  // global strings in LLVM-IR, which can be caused by multiple translation of
+  // UserSemantic decorations with the same literal. This map is used to store
+  // global strings by their annotations.
+  std::unordered_map<std::string, Constant *> AnnotationsMap;
```
Comment (mlychkov) [inline lib/SPIRV/SPIRVReader.h:170]:
```suggestion
  // AnnotationsMap helps to translate annotation strings for local variables.
  // Map values are pointers on global strings in LLVM-IR.
  // It is used to avoid duplication of these annotation strings in LLVM-IR,
  // which can be caused by multiple translation of
  // UserSemantic decorations with the same literal.
  std::unordered_map<std::string, Constant *> AnnotationsMap;
```

File: lib/SPIRV/SPIRVReader.h (hunk 5 above)
Comment (KornevNikita) [reply lib/SPIRV/SPIRVReader.h:164]: Done

**Review by mlychkov — APPROVED** [review]
LGTM, thank you.
